vxtelegram/telegram.py

```diff
@@ -157,31 +185,104 @@ def handle_outbound_message(self, message):
         url = self.get_outbound_url('sendMessage')
         http_client = HTTPClient(self.agent_factory())
 
-        try:
-            r = yield http_client.post(
-                url=url,
-                data=json.dumps(outbound_msg),
-                headers={'Content-Type': ['application/json']},
-            )
+        r = yield http_client.post(
+            url=url,
+            data=json.dumps(outbound_msg),
+            headers={'Content-Type': ['application/json']},
+            allow_redirects=False,
+        )
+
         # Telegram redirects our request if our bot token is invalid
-        except ResponseFailed as e:
-            yield self.outbound_failure(message_id,
-                                        'Invalid token (redirected)\n%s' % e)
+        if r.code == http.FOUND:
```

Comment:
@justinvdm it's annoying that they don't actually document how they respond to invalid requests, but it looks like they evaluate the token against a regular expression. Using a token like '1234' returns a 302, but using a valid token with just one symbol removed returns a 401 unauthourised with a proper JSON body (unless you remove the underscore or colon, then you get a 302 as well). So it seems like 302 means "this isn't a token, we assume you're just looking for the bot API docs", and 401 means "this could be a token, but it's not registered with us".

Other than that, we get a 502 if we don't specify a method path, a 404 if the method doesn't exist ('sendMessige'), a 400 if our params are invalid, and a 401 if we attempt to send a message to a chat/user that's been deleted.